prov/sockets/src/sock_rx_entry.c

```diff
@@ -49,30 +49,54 @@
 #define SOCK_LOG_DBG(...) _SOCK_LOG_DBG(FI_LOG_EP_DATA, __VA_ARGS__)
 #define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
 
-/* FIXME: pool of rx_entry */
 struct sock_rx_entry *sock_rx_new_entry(struct sock_rx_ctx *rx_ctx)
 {
 	struct sock_rx_entry *rx_entry;
+	struct dlist_entry *entry;
+	int i;
 
-	rx_entry = calloc(1, sizeof(*rx_entry));
-	if (!rx_entry)
-		return NULL;
+	if (rx_ctx->rx_entry_pool == NULL) {
+		rx_ctx->rx_entry_pool = calloc(SOCK_DEF_RX_POOL_SIZE, sizeof (*rx_entry));
+		dlist_init(&rx_ctx->pool_list);
+		
+		for (i = 0; i < SOCK_DEF_RX_POOL_SIZE; i++) {
+			dlist_insert_tail(&rx_ctx->rx_entry_pool[i].entry,
+					  &rx_ctx->pool_list);
+			rx_ctx->rx_entry_pool[i].is_pool_entry = 1;
+		}
+	}
+
+	if (!dlist_empty(&rx_ctx->pool_list)) {
+		entry = rx_ctx->pool_list.next;
+		rx_entry = container_of(entry, struct sock_rx_entry, entry);
+		rx_entry->rx_ctx = rx_ctx;
+		dlist_remove(entry);
+	} else {
+		rx_entry = calloc(1, sizeof(*rx_entry));
+		if (!rx_entry)
+			return NULL;
+	}
```

Comment:
I don't see anything tracking these allocations, so that they can be freed later.

If this is a free pool, using a single linked list as a queue will be better performing.  Removing an object from a double linked list is expensive by comparison.

There is likely already code that handles a free pool of objects in the tree.  @j-xiong - I believe the PSM has this as well.  We should move this into a common area.